command/auto_complete.go

```diff
@@ -0,0 +1,247 @@
+`
+
+const bash = `# prepare autocompletion suggestions for s5cmd and save them to COMPREPLY array
+_s5cmd_cli_bash_autocomplete() {
+	if [[ "${COMP_WORDS[0]}" != "source" ]]; then
+		COMPREPLY=()
+		local opts cur cmd` +
+	// get current word (cur) and prepare command (cmd)
+	`
+		cur="${COMP_WORDS[COMP_CWORD]}"
+		cmd="${COMP_LINE:0:$COMP_POINT}"` +
+
+	// if we want to complete the second argurment and we didn't started writing
+	// yet then we should pass an empty string as another argument. Otherwise
+	// the white spaces will be discarded and the program will make suggestions
+	// as if it is completing the first argument.
+	// Beware that the we want to pass empty string so we intentionally write
+	// as it is. Fixes of SC2089 and SC2090 are not what we want.
+	// see also https://www.shellcheck.net/wiki/SC2090
+	`
+		[ "${COMP_LINE:COMP_POINT-1:$COMP_POINT}" == " " ] \
+			&& cmd="${cmd} \"\"" ` +
+
+	// execute the command with '--generate-bash-completion' flag to obtain
+	// possible completion values for current word.
+	// ps. SC2090 is not wanted.
+	`
+		opts=$($cmd --generate-bash-completion)` +
+
+	// prepare completion array with possible values and filter those do not start with cur.
+	// if no completion is found then fallback to default completion of shell.
+	`
+
+		while IFS='' read -r line; do COMPREPLY+=("$line"); done < <(compgen -o bashdefault -o default -o nospace -W "${opts}" -- "${cur}")
+
+		return 0
+	fi
+}
+
+# call the _s5cmd_cli_bash_autocomplete to complete s5cmd command. 
+complete -o nospace -F _s5cmd_cli_bash_autocomplete s5cmd
+`
+
+const pwsh = `$fn = $($MyInvocation.MyCommand.Name)
+$name = $fn -replace "(.*)\.ps1$", '$1'
+Register-ArgumentCompleter -Native -CommandName $name -ScriptBlock {
+	param($commandName, $wordToComplete, $cursorPosition)
+	$other = "$wordToComplete --generate-bash-completion"
+		Invoke-Expression $other | ForEach-Object {
+		[System.Management.Automation.CompletionResult]::new($_, $_, 'ParameterValue', $_)
+		}
+}
+`
+
+func getBashCompleteFn(cmd *cli.Command, isOnlyRemote, isOnlyBucket bool) func(ctx *cli.Context) {
+	isOnlyRemote = isOnlyRemote || isOnlyBucket
+	return func(ctx *cli.Context) {
+		var arg string
+		args := ctx.Args()
+		l := args.Len()
+
+		if l > 0 {
+			arg = args.Get(l - 1)
+		}
+
+		// argument may start with a quotation mark, in this case we want to trim that before
+		// checking if it has prefix s3://
+		// Beware that we only want to trim the first char, not all of the leading
+		// quotation marks, because those quotation marks may be actual charactes.
+		if strings.HasPrefix(arg, "'") {
+			arg = strings.TrimPrefix(arg, "'")
+		} else {
+			arg = strings.TrimPrefix(arg, "\"")
+		}
+
+		if isOnlyRemote || strings.HasPrefix(arg, "s3://") {
+			if strings.HasPrefix(arg, "-") {
+				cli.DefaultCompleteWithFlags(cmd)(ctx)
+				return
+			}
+			printS3Suggestions(ctx, arg, isOnlyBucket)
+			return
+		}
+
+		cli.DefaultCompleteWithFlags(cmd)(ctx)
+	}
+}
+
```

Comment:
With the comment above making these functions a method of a new type, what do you think writing individual tests for these? They're quite tricky to grasp.

Reply:
Maybe separating logic for `bash` from others -in all autocomplete functions- would help as it is the one that requires tinkering